man/man3/wtt.3
.\" (c) 2001 by Poonlap Veerathanabutr (Poonlap.Veerathanabutr@sun.co.jp)
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\" License.
.TH wtt 3  "Sep 16, 2001" "Thai Linux Working Group" "libthai's Manual"
.SH NAME
\fBwtt\fR \- WTT I/O implementation
.sp
\fBTAC_chtype\fR \- Determine type of a character defined by TACTIS.
.br
\fBTACio_op \- Check composibility of two chracters.
.SH SYNOPSIS
\fB#include <thai/wtt.h>\fR
.br
\fIint\fR \fBTACchtype\fR(\fIthchar_t c\fR);
.br
\fIint\fR \fBTACio_op\fR(\fIthchar_t c1\fR, \fIthchar_t c2\fR);
.br
.SH DESCRIPTION
WTT stands for Wing Tuk Tee (in Thai, Runs everywhere). It was defined by TACTIS (Thai API Consortium/Thai Industrial Standard) in the NECTEC Thai Software Standard Project (1989-1991), and later endorsed by Thai Industrial Standard Institute (TISI) as TIS 1566-2541 in 1998.
.PP
WTT classifies Thai chracter(TIS-620) into 17 types below.
.sp
\fITYPE  VALUE SHORT_DESCRIPTION\fR
.br
CTRL    0    control characters
.br
NON     1    non composible characters
.br
CONS    2    consonants
.br
LV      3    leading vowels
.br
FV1     4    following vowels 1
.br
FV2     5    following vowels 2
.br
FV3     6    following vowels 3
.br
BV1     7    below vowels 1
.br
BV2     8    below vowels 2
.br
BD      9    below diacritics
.br
TONE    10   tonemarks
.br
AD1     11   above diacritics 1
.br
AD2     12   above diacritics 2
.br
AD3     13   above diacritics 3
.br
AV1     14   above vowels 1
.br
AV2     15   above vowels 2
.br
AV3     16   above vowels 3
.sp
Functions in \fBthctype\fR do basic character classifications while \fBwtt\fR classifies a chracter in detail. Please refer to the reference.
.sp
\fBTACio_op\fR checks how to compose two given chracters. The possiblities are Composible(CP), Non-display(XC), Accept(AC), Reject(RJ) and Strict Reject(SR). The values of CP, XC, AC, RJ and SR are difined in wtt.h. And their meanings are:
.sp
CP : second character is displayed in the same cell as the first, also implies an acceptance. 
.br
XC : Do nothing.
.br
AC : Display second character in the next cell.
.br
RJ : Discard second character.
.br
SR : Reject second character only in strict mode.
.SH RETURN VALUE
\fBTACchtype\fR returns an interger defined in the file wtt.h.
.br
\fBTACio_op\fR returns CP, XC, AC, RJ or SR.
.SH "SEE ALSO"
.B libthai(3), thctype(3)
.br
"Thai Input and Output Methods" by Dr.Thaweesak Koanantokool
.SH AUTHORS
\fBProject Leader\fR
.br
Theppitak Karoonboonyanan <theppitak@gmail.com>
.br
\fBMembers\fR
.br
Chanop Silpa-Anan <chanop@syseng.anu.edu.au>
.br
Pattara Kiatisevi <ott@linux.thai.net>
.br
Vuthichai Ampornaramveth <vuthi@nii.ac.jp>
.br
Poonlap Veerathanabutr <Poonlap.Veerathanabutr@sun.co.jp>


NEWS
=======================================================================
RELEASES INFORMATION
=======================================================================

LibThai 0.1.21 (21 August 2014):
* Updated word break dictionary.
* Improve word breaking performance by 8.61%

LibThai 0.1.20 (25 October 2013):
* Updated word break dictionary.
* Fix compiler warnings in test suites.

LibThai 0.1.19 (29 January 2013):
* New API th_validate_leveled() for IM correction with strictness level.
* Updated word break dictionary.
* Switched to XZ tarball compression.

LibThai 0.1.18 (12 June 2012):
* Updated word break dictionary.

LibThai 0.1.17 (21 February 2012):
* Recognizes acronyms on word breaking.
  (Thanks John Tapsell for the report and initial patch.)
* Updated word break dictionary.

LibThai 0.1.16 (5 November 2011):
* Updated word break dictionary.

LibThai 0.1.15 (24 March 2011):
* Updated word break dictionary.

LibThai 0.1.14 (28 February 2010):
* Fix crash on corrupted dictionary file; now it fails quietly instead.
* License clarification on individual source files.
* Updated word break dictionary.

LibThai 0.1.13 (15 January 2010):
* Fix integer overflow vulnerabilities (CVE-2009-4012).
* Updated word break dictionary.

LibThai 0.1.12 (18 June 2009):
* Improved unknown words boundary.
* Add symbol versioning.
* Updated word break dictionary.

LibThai 0.1.11 (6 April 2009):
* Diminished link flags in pkg-config info, for less dependency for clients.
* More words in word break dictionary.

LibThai 0.1.10 (30 March 2009):
* More dictionary capacity with libdatrie 0.2.
* Dictionary clean-up on unload.
* Some performance fine-tuning.
* More words in word break dictionary.

LibThai 0.1.9 (28 August 2007):
* Rewritten itemization for line breaker, for conformance to UAX#14.
* Revised dictionary for compound words support.

LibThai 0.1.8 (3 March 2007):
* Improved handling of white spaces in word break routine, for better quality
  line wrapping in Gecko libthai component.
* More words in word break dictionary.
* Excessive exported symbols are eliminated from library.

LibThai 0.1.7 (14 October 2006):
* Rewritten word break routines:
  - Split dictionary from source, thus new dependency on libdatrie
  - New word break routine from scratch, for maintainability
* Improved dictionary for word segmentation:
  - New word list preparation process from multiple categories
  - New contemporary words not included in Royal Institute dictionary,
    as well as some proper names for countries and locations, etc.
  - Fine-tuned dictionary to get rid of some potential ambiguities

LibThai 0.1.6 (5 August 2006):
* Documentation with doxygen.
* Add missing APIs.
* Fix compiler warnings.

LibThai 0.1.5 (13 March 2006):
* Improved word breaking. Space is now also considered a breakable point.
* More PUA support for Mac fonts rendering.
* Start maintaining library version info from now on.

LibThai 0.1.4 (16 October 2004):

* Handle the rendering of NIKHAHIT and MAITAIKHU for Pali and Kuy,
  respectively, in which the marks can be in either above or top level.
* Input sequence validation for NIKHAHIT and MAITAIKHU is improved as well.
* Render the error case of repeated SARA AM's more gracefully.

LibThai 0.1.3 (22 February 2004):

* Define blank base glyph for floating upper/lower vowels rendering.
* (Minor modified version as required by pango-libthai binding).

LibThai 0.1.2 (10 September 2003):

* It's still alive!
* Bug fix & code furnishing
* conversion functions for Win & Mac extension
* Some document (man pages)
* Experiment with GTK+ 2 plug-ins (in separate packages)

LibThai 0.1.1 (14 August 2001):

* WTT-based cell clustering
* Thai input sequence correction
* Thai shaping

LibThai 0.1.0 (4 August 2001):

* First public release, request for comments for API
* Thai character properties with WTT 2.0 tables
* Thai string collation (bilingual)
* Thai string normalization
* WTT-based input sequence checking
* Thai Word breaking (cttex)
* Conversion between TIS-620 and Unicode
* All implementations not yet completed

=======================================================================
See ChangeLog file for more details.



src/thctype/wtt.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * libthai - Thai Language Support Library
 * Copyright (C) 2001  Theppitak Karoonboonyanan <theppitak@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

/*
 * wtt.c - WTT I/O implementation
 * Created: 2001-08-04
 * Author:  Theppitak Karoonboonyanan <theppitak@gmail.com>
 */

#include <thai/wtt.h>

short TACchtype_[256] = {
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /*  0 -  7 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /*  8 - 15 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /* 16 - 23 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /* 24 - 31 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 32 - 39 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 40 - 47 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 48 - 55 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 56 - 63 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 64 - 71 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 72 - 79 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 80 - 87 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 88 - 95 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 96 - 103 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 104 - 111 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 112 - 119 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  CTRL,  /* 120 - 127 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /* 128 - 135 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /* 136 - 143 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /* 144 - 151 */
    CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL, CTRL,  /* 152 - 159 */
    NON,  CONS, CONS, CONS, CONS, CONS, CONS, CONS,  /* 160 - 167 */
    CONS, CONS, CONS, CONS, CONS, CONS, CONS, CONS,  /* 168 - 175 */
    CONS, CONS, CONS, CONS, CONS, CONS, CONS, CONS,  /* 176 - 183 */
    CONS, CONS, CONS, CONS, CONS, CONS, CONS, CONS,  /* 184 - 191 */
    CONS, CONS, CONS, CONS,  FV3, CONS,  FV3, CONS,  /* 192 - 199 */
    CONS, CONS, CONS, CONS, CONS, CONS, CONS, NON,   /* 200 - 207 */
    FV1,  AV2,  FV1,  FV1,  AV1,  AV3,  AV2,  AV3,   /* 208 - 215 */
    BV1,  BV2,  BD,   NON,  NON,  NON,  NON,  NON,   /* 216 - 223 */
    LV,   LV,   LV,   LV,   LV,   FV2,  NON,  AD2,   /* 224 - 231 */
    TONE, TONE, TONE, TONE, AD1,  AD1,  AD3,  NON,   /* 232 - 239 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  NON,   /* 240 - 247 */
    NON,  NON,  NON,  NON,  NON,  NON,  NON,  CTRL   /* 248 - 255 */
};

/* Table for Thai Cell Manipulation */
short TACio_op_[17][17] = {
      /* Table 2: WTT I/O sequence check rules */
      /* row: leading char,  column: following char */
/* CTRL NON CONS LV FV1 FV2 FV3 BV1 BV2 BD TONE AD1 AD2 AD3 AV1 AV2 AV3 */
   {XC, AC, AC, AC, AC, AC, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*CTRL*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*NON*/
  ,{XC, AC, AC, AC, AC, SR, AC, CP, CP, CP, CP, CP, CP, CP, CP, CP, CP}/*CONS*/
  ,{XC, SR, AC, SR, SR, SR, SR, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*LV*/
  ,{XC, AC, AC, AC, AC, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*FV1*/
  ,{XC, AC, AC, AC, AC, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*FV2*/
  ,{XC, AC, AC, AC, SR, AC, SR, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*FV3*/
  ,{XC, AC, AC, AC, AC, SR, AC, RJ, RJ, RJ, CP, CP, RJ, RJ, RJ, RJ, RJ}/*BV1*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, CP, RJ, RJ, RJ, RJ, RJ, RJ}/*BV2*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*BD*/
  ,{XC, AC, AC, AC, AC, AC, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*TONE*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*AD1*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*AD2*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ, RJ}/*AD3*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, CP, CP, RJ, RJ, RJ, RJ, RJ}/*AV1*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, CP, RJ, RJ, RJ, RJ, RJ, RJ}/*AV2*/
  ,{XC, AC, AC, AC, SR, SR, AC, RJ, RJ, RJ, CP, RJ, CP, RJ, RJ, RJ, RJ}/*AV3*/
};

#undef TACchtype
#undef TACio_op

/**
 * @brief  WTT character class
 * @param  c : the character
 * @return  WTT character class { CTRL, ..., AV3 }
 */
WTTClass TACchtype(thchar_t c)
{
    return TACchtype_[c];
}

/**
 * @brief  WTT I/O operation
 * @param  c1 : previous character
 * @param  c2 : next character
 * @return  WTT I/O operation { CP, XC, AC, RJ, SR }
 */
WTTOp    TACio_op(thchar_t c1, thchar_t c2)
{
    return TACio_op_[TACchtype_[c1]][TACchtype_[c2]];
}

/*
vi:ts=4:ai:expandtab
*/
